src/communication/websocket/auth.hpp

```diff
@@ -19,15 +19,24 @@
 
 namespace communication::websocket {
 
-class SafeAuth {
+class IAuthentication {
+ public:
+  virtual bool Authenticate(const std::string &username, const std::string &password) const = 0;
+
+  virtual bool HasUserPermission(const std::string &username, auth::Permission permission) const = 0;
+
+  virtual bool HasAnyUsers() const = 0;
+};
+
+class SafeAuth : public IAuthentication {
```

Comment:
Why not `CRTP` here? You don't really need runtime polymorphism here.

Reply:
I guess the only reason I would not prefer CRTP here is that I would have to have only header files for the listener and session. And I don't think that inheritance here will ever be a bottleneck since authentication is a straightforward process. What do you think about that?

Reply:
personally, I think if something can be expressed at compile time it should be done at compile time. However, that's idealistic and it's not going to make you more productive nor will it help you on this PR. I agree it won't be a bottleneck so plz feel free to keep things as is